CentaurusInfra/arktos
# [Mizar-Arktos-Integration]Tenant-vpc-support: code changes for mizar pod controller and mizar arktos network controller

**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:
In terms of the design document at https://github.com/CentaurusInfra/arktos/blob/poc-2022-01-30/docs/design-proposals/multi-tenancy/multi-tenancy-network-mizar-integration.md, this PR implements the following two functions:

1. Mizar-arktos-network-controller (existing controller) automatically creates Mizar CRDs - VPC and subnet, based on vpcID coming from corresponding default arktos network object, which is in Ready state, using Dynamic Client technology including Discovery REST mapper and Dynamic REST interface.

The json template files for Mizar CRDs - VPC and Subnet are located as default at the directory ./hack/runtime/.

./hack/runtime/default_mizar_network_subnet_template.json (the below is an example)
```
{
   "apiVersion": "mizar.com/v1",
   "kind": "Vpc",
   "metadata": {
      "name": "{{.Tenant}}-default-network"
   },
   "spec": {
      "ip": "21.0.0.0",
      "prefix": "16",
      "dividers": 1,
      "status": "Init"
   }
}
```
./hack/runtime/default_mizar_network_vpc_template.json (the below is an example)
```
{
   "apiVersion": "mizar.com/v1",
   "kind": "Subnet",
   "metadata": {
      "name": "{{.Tenant}}-default-network-subnet"
   },
   "spec": {
      "ip": "21.0.21.0",
      "prefix": "24",
      "bouncers": 1,
      "vpc": "{{.Tenant}}-default-network",
      "status": "Init"
   }
}
```

2. Mizar-pod-controller (existing controller) gets vpcID information from arktos network object in Ready state, automatically creates the following annotations, then patches the objects(pods) with correct annotations and updates to API server, finally notifies the annotations to Mizar via grpcAdaptor.

```
    mizar.com/subnet: <tenant name>-default-network-subnet
    mizar.com/vpc: <tenant name>-default-network
```

**Which issue(s) this PR fixes**:
The new feature is part of Mizar Arktos integration.

Fixes # N/A

**Special notes for your reviewer**:
Test steps:

Note: These steps have been tested with system tenant in Carl's local dev environment which runs the codes in poc-2022-01-30 branch q131172019:CarlXie-poc-2022-01-30-tenant-vpc-support,

1. Configure two template files of vpc and subnet
    1) ./hack/runtime/default_mizar_network_subnet_template.json
        - change the parameters - **ip, prefix, dividers** based on above example
     2) ./hack/runtime/default_mizar_network_vpc_template.json
        - change the parameters - **ip, prefix, bouncers** based on above example

2. Run command "**CNIPLUGIN=mizar ./hack/arktos-up.sh**" to start Arktos scale-up cluster with Mizar. Or follow up the steps 0.6) through step 1 at https://github.com/q131172019/arktos/blob/CarlXie_singleNodeArktosCluster/docs/setup-guide/single-node-dev-scale-up-cluster-with-Mizar.md to ensure Mizar CRDs and pods are in Running state, especially check whether CRDs - **vpc (system-default-network) and subnet (system-default-network-subnet)** as well as dividers and bouncers of tenant 'system' are created correctly based on above two template files of VPC and Subnet .

3. Use the following yaml file to create two nginx pods of tenant 'system'
~/TMP/mizar/pod-vpc-1-without-annotation.yaml
```
apiVersion: v1
kind: Pod
metadata:
  name: ying-nginx-system-1
  tenant: system
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
      - containerPort: 443
---
apiVersion: v1
kind: Pod
metadata:
  name: ying-nginx-system-2
  tenant: system
spec:
  containers:
  - name: nginx
    image: nginx
    ports:
      - containerPort: 443
```
```
./cluster/kubectl.sh apply -f ~/TMP/mizar/pod-vpc-1-without-annotation.yaml
```
4. Check whether the annotations including two mizar CRDs - vpc and subnet are correctly added into two nginx pods
```
./cluster/kubectl.sh get pods/ying-nginx-system-1 --tenant system -o yaml |more
./cluster/kubectl.sh get pods/ying-nginx-system-2 --tenant system -o yaml |more
```
```
  annotations:
    <snip> 
    mizar.com/subnet: system-default-network-subnet
    mizar.com/vpc: system-default-network
```
5. Check whether two nginx pods can communicate each other via curl
```
POD1="ying-nginx-system-1"
POD2="ying-nginx-system-2"
IP_POD1=`./cluster/kubectl.sh get pods --tenant system -o wide |grep $POD1 |awk '{print $7}'`
IP_POD2=`./cluster/kubectl.sh get pods --tenant system -o wide |grep $POD2 |awk '{print $7}'`
echo "$POD1: $IP_POD1"
echo "$POD2: $IP_POD2"
./cluster/kubectl.sh exec -ti $POD1 --tenant system -- curl $IP_POD2
./cluster/kubectl.sh exec -ti $POD2 --tenant system -- curl $IP_POD1
```

6. Delete nginx pods after verification
```
./cluster/kubectl.sh delete -f ~/TMP/mizar/pod-vpc-1-without-annotation.yaml
```

**Does this PR introduce a user-facing change?**:
No.
